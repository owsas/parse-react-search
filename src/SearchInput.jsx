import React from 'react';
import PropTypes from 'prop-types';
import Autocomplete from 'react-autocomplete';
import CheckCircleIcon from 'react-icons/lib/fa/check-circle-o';

export default class SearchInput extends React.Component {
  /**
   * Given a country code, brings the url to the flag in aws
   * @param {string} code
   */
  static getCountryFlagURL(code) {
    return `https://s3-us-west-2.amazonaws.com/otherwise-files/flags/${code.toLowerCase()}.png`;
  }

  constructor(props) {
    super(props);

    this.state = {
      value: props.initialValues.value,
      results: [],
      query: this.props.query,
      querySearchKey: this.props.querySearchKey,
      transformItem: this.props.transformItem,
    };

    this.selected = props.initialValues.selected;
    this.timeout = undefined;

    this.onChange = this.onChange.bind(this);
    this.onSelect = this.onSelect.bind(this);
    this.renderItem = this.renderItem.bind(this);
    this.clearValue = this.clearValue.bind(this);
  }

  componentWillReceiveProps(props = {}) {
    if (props.initialValues && props.initialValues.value) {
      this.setState({ value: props.initialValues.value });
      this.selected = props.initialValues.selected;
    }

    if (props.query) {
      this.setState({
        query: props.query,
      });
    }

    if (props.transformItem) {
      this.setState({
        transformItem: props.transformItem,
      });
    }

    if (props.querySearchKey) {
      this.setState({
        querySearchKey: props.querySearchKey,
      });
    }
  }

  onChange(e) {
    this.setState({ value: e.target.value });
    this.selected = undefined;

    if (this.timeout) {
      window.clearTimeout(this.timeout);
      this.timeout = undefined;
    }

    this.timeout = window.setTimeout(() => {
      this.findResults();
    }, 300);
  }

  onSelect(value, item) {
    this.selected = item;
    this.setState({ value });

    if (this.props.onSelect) {
      this.props.onSelect(item);
    }
  }

  getSelectedItem() {
    return this.selected;
  }

  getValue() {
    return this.getSelectedItem() || { text: this.state.value };
  }

  clearValue() {
    this.setState({ value: '' });
  }

  async findResults() {
    try {
      const {
        queryOptions,
        includeSearchText,
      } = this.props;

      const {
        query,
        querySearchKey,
        transformItem,
      } = this.state;

      let results;

      if (query && transformItem) {
        if (querySearchKey) {
          query.matches(querySearchKey, this.state.value);
        }

        results = await query.find(queryOptions).then(transformItem);
      } else {
        results = []; // TODO: Call ParseReactSearch
      }

      if (includeSearchText) {
        const localResults = this.state.value ? [{ text: this.state.value }] : [];

        this.setState({ results: [].concat(localResults, results) });
      } else {
        this.setState({ results });
      }
    } catch (e) {
      // nothing
    }
  }

  renderItem(item, isHighlighted) {
    if (!item) return null;

    return (
      <div key={`searchInput-item-${item.objectId}`} style={{ background: isHighlighted ? 'lightgray' : 'white', padding: 5 }}>
        <div style={{ display: 'flex', flexDirection: 'row', alignItems: 'center' }}>
          {item.picUrl &&
            <div>
              <img src={item.picUrl} alt="" style={{ height: 40, width: 40 }} />
            </div>
          }
          <div style={{ flex: 1, padding: 5 }}>
            <p style={{ margin: 0 }}><b>{item.text}</b> {item.official && <CheckCircleIcon />}</p>
            <p style={{ margin: 0 }}>
              {item.countryCode &&
                <img
                  src={SearchInput.getCountryFlagURL(item.countryCode)}
                  style={{ width: 16, marginRight: 5 }}
                  alt=""
                />
              }
              {item.alias && <span>{item.alias}</span>}
              {item.address && <span>{item.address}</span>}
            </p>
          </div>
        </div>
      </div>
    );
  }

  render() {
    const { value, results } = this.state;

    return (
      <Autocomplete
        {...this.props.autocompleteProps}
        getItemValue={item => item.text}
        items={results}
        renderItem={this.renderItem}
        value={value}
        onChange={this.onChange}
        onSelect={this.onSelect}
        wrapperStyle={{
          position: 'relative',
        }}
      />
    );
  }
}

SearchInput.defaultProps = {
  initialValues: {},
  query: undefined,
  queryOptions: undefined,
  transformItem: undefined,
  scope: ['Brand'],
  autocompleteProps: {},
  includeSearchText: true,
  onSelect: undefined,
  querySearchKey: undefined,
};

SearchInput.propTypes = {
  initialValues: PropTypes.shape({
    value: PropTypes.string,
    selected: PropTypes.object, // eslint-disable-line
  }),
  query: PropTypes.object, // eslint-disable-line
  queryOptions: PropTypes.object, // eslint-disable-line
  transformItem: PropTypes.func,
  scope: PropTypes.arrayOf(PropTypes.string),
  includeSearchText: PropTypes.bool,
  autocompleteProps: PropTypes.object, //eslint-disable-line
  onSelect: PropTypes.func,
  querySearchKey: PropTypes.string,
};
