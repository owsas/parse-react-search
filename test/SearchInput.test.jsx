import React from 'react';
import { mount, shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import SearchInput from '../src/SearchInput';

describe('#getCountryFlagURL', () => {
  test('should convert upper case to lowercase', () => {
    const countryCode = 'US';
    const expected = `https://s3-us-west-2.amazonaws.com/otherwise-files/flags/${countryCode.toLowerCase()}.png`;

    const result = SearchInput.getCountryFlagURL(countryCode);
    expect(result).toEqual(expected);
  });
});

describe('#getValue', () => {
  test('should get the value on the state', () => {
    const m = mount(<SearchInput />);
    m.setState({ value: 'test value' });
    expect(m.instance().getValue()).toEqual({ text: 'test value' });
  });

  test('should return the selected item if it exists', () => {
    const m = mount(<SearchInput />);
    m.setState({ value: 'test value' });
    m.instance().selected = { test: 'ok' };
    expect(m.instance().getValue()).toEqual({ test: 'ok' });
  });
});

describe('#renderItem', () => {
  test('should match the snapshot', () => {
    const m = shallow(<SearchInput />);
    const m2 = shallow(m.instance().renderItem({
      text: 'Test',
      countryCode: 'CO',
      objectId: '123',
      official: true,
      picUrl: 'testUrl',
    }));
    expect(toJSON(m2)).toMatchSnapshot();
  });
});

test('should match the snapshot', () => {
  const m = shallow(<SearchInput />);
  expect(toJSON(m)).toMatchSnapshot();
});
