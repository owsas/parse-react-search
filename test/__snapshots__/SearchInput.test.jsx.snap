// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#renderItem should match the snapshot 1`] = `
<div
  key="searchInput-item-123"
  style={
    Object {
      "background": "white",
      "padding": 5,
    }
  }
>
  <div
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "row",
      }
    }
  >
    <div>
      <img
        alt=""
        src="testUrl"
        style={
          Object {
            "height": 40,
            "width": 40,
          }
        }
      />
    </div>
    <div
      style={
        Object {
          "flex": 1,
          "padding": 5,
        }
      }
    >
      <p
        style={
          Object {
            "margin": 0,
          }
        }
      >
        <b>
          Test
        </b>
         
        <FaCheckCircleO />
      </p>
      <p
        style={
          Object {
            "margin": 0,
          }
        }
      >
        <img
          alt=""
          src="https://s3-us-west-2.amazonaws.com/otherwise-files/flags/co.png"
          style={
            Object {
              "marginRight": 5,
              "width": 16,
            }
          }
        />
      </p>
    </div>
  </div>
</div>
`;

exports[`should match the snapshot 1`] = `
<Autocomplete
  autoHighlight={true}
  getItemValue={[Function]}
  inputProps={Object {}}
  isItemSelectable={[Function]}
  items={Array []}
  menuStyle={
    Object {
      "background": "rgba(255, 255, 255, 0.9)",
      "borderRadius": "3px",
      "boxShadow": "0 2px 12px rgba(0, 0, 0, 0.1)",
      "fontSize": "90%",
      "maxHeight": "50%",
      "overflow": "auto",
      "padding": "2px 0",
      "position": "fixed",
    }
  }
  onChange={[Function]}
  onMenuVisibilityChange={[Function]}
  onSelect={[Function]}
  renderInput={[Function]}
  renderItem={[Function]}
  renderMenu={[Function]}
  selectOnBlur={false}
  value=""
  wrapperProps={Object {}}
  wrapperStyle={
    Object {
      "position": "relative",
    }
  }
/>
`;
